pkg/cloudprovider/aws/apis/v1alpha1/provider.go

```diff
@@ -74,6 +74,25 @@ type LaunchTemplate struct {
 	// BlockDeviceMappings to be applied to provisioned nodes.
 	// +optionals
 	BlockDeviceMappings []*BlockDeviceMapping `json:"blockDeviceMappings,omitempty"`
+	// UserData to be applied to the provisioned nodes.
+	// It must be in the appropriate format based on the specified AMIFamily. Karpenter will merge certain fields into
+	// this UserData to ensure nodes are being provisioned with the correct configuration.
+	// +optional
+	UserData *UserData `json:"userData,omitempty"`
```

Comment:
Did we discuss the tradeoffs of doing this inline vs as a configmap (e.g. the design doc?). I worry a bit about resource sprawl. One huge advantage I can see is that separating this into its own resource allows users to configure the same user data for multiple provisioners. 

In #1327, we propose separating the entire Provider object into its own CRD. At that point, the AWS Provider is more or less a kube-native launchTemplate-like abstraction, and shares the benefit mentioned above about being used by multiple provisioners.

Reply:
@bwagner5 and I have been discussing this offline for a while (and in this PR too), however it's not captured in the design doc. 

This PR started off by inlining the UserData but it made the Provisioner much too large and quite unreadable. UserData can be 16KB of data, and as I started to play with it more it makes a lot of sense to keep it outside. For example, when using this with AL2, it was really odd looking to have bash scripts inlined inside the provisionerSpec so we decided to separate that out. 


> One huge advantage I can see is that separating this into its own resource allows users to configure the same user data for multiple provisioners.

Completely agree, and that's another reason why we think this approach is better than the inlining. 

> In #1327, we propose separating the entire Provider object into its own CRD. At that point, an AWSProvider CRD is more or less a kube-native launch-template abstraction, and shares the benefit mentioned above about being used by multiple provisioners.

I think I made this exact comment yesterday when demo-ing some of this stuff, and I 100% agree :). The ConfigMap solution is better than inlining, but doesn't go as far as it should. A dedicated CRD that's essentially an EC2 launch template should be next.

Reply:
I might start by inlining it in the provisioner and then migrate to a separate CRD in the future. This is a more natural concept progression, since if it's eventually it's own CRD, it will be alongside other provider fields. This also seems like an easier migration path for customers, since they don't have to worry about cleaning up configmaps from a previous version.

Reply:
The demo @suket22 gave used the [configmap from file](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-files) feature which made for a nice way to create the userdata.  This only works with configmaps as far as I can tell.

Reply:
Can you paste an example of an unreadable provisioner spec with it inlined? Id image we'd use yaml multi line strings and match indentation. I'd also expect that users provide this at the bottom.

Reply:
Something like this for AL2 - 
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  limits:
    resources:
      cpu: 20
  provider:
    amiFamily: AL2
    userData:  |
      MIME-Version: 1.0
      Content-Type: multipart/mixed; boundary="//"

      --//
      Content-Type: text/x-shellscript; charset="us-ascii"

      #!/bin/bash
      echo "Running custom user data script"

      --//--      
    instanceProfile: KarpenterNodeInstanceProfile-karp-cluster
    subnetSelector:
      karpenter.sh/discovery: karp-cluster
    securityGroupSelector:
      karpenter.sh/discovery: karp-cluster
  ttlSecondsAfterEmpty: 30
```

I'd expect most user data scripts to run *far longer* than that however, making it odd to manage IMO.

Reply:
Does it look cleaner if userdata is less nested (i.e. its own CRD)? What if you do it as the final key in the struct (e.g. alpha order)? Is it possible to default the mime parts to shell script (the way LTs allow today) to remove some of the cruft?

Reply:
I think it does make it look cleaner if it's not inlined. This is what it looks like if it's a ConfigMap instead
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  limits:
    resources:
      cpu: 20
  provider:
    amiFamily: AL2
    userData:
      configMap:
        name: al2-config
        namespace: karpenter
    instanceProfile: KarpenterNodeInstanceProfile-karp-cluster
    subnetSelector:
      karpenter.sh/discovery: karp-cluster
    securityGroupSelector:
      karpenter.sh/discovery: karp-cluster
  ttlSecondsAfterEmpty: 30
```

> Is it possible to default the mime parts to shell script

I don't think we can assume that though because someone might give Karpenter a UserData composed of multiple MIME parts already.

Reply:
This is what I was getting at:
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  limits:
    resources:
      cpu: 20
  ttlSecondsAfterEmpty: 30
  provider:
    amiFamily: AL2
    instanceProfile: KarpenterNodeInstanceProfile-karp-cluster
    subnetSelector:
      karpenter.sh/discovery: karp-cluster
    securityGroupSelector:
      karpenter.sh/discovery: karp-cluster
    userData:  |
      MIME-Version: 1.0
      Content-Type: multipart/mixed; boundary="//"
      --//
      Content-Type: text/x-shellscript; charset="us-ascii"
      #!/bin/bash
      echo "Running custom user data script"
      --//--      
```
If you format the user data at the bottom, I'm not sure it's so bad. 

I think the namespacing of a configmap is a showstopper since provisioners aren't namespaced. I wouldn't be able to recommend a good namespace to customers. Customers may choose `default` or `karpenter`, but both of these are bad choices (happy to elaborate if this is unintuitive to you). We then end up down the road of recommending customers create a namespace like `provisioner-userdata`, but this may be challenging to setup, since many clusters have tight control over namespaces -- this has security implications.

Minimally, if we do go the route of what you're proposing, does it make sense to do something like `userDataConfigMap` instead of `userData.ConfigMap`, since I don't think userData will ever have additional fields.

Reply:
w/ v1alpha6, we'd end up with something like this
```
apiVersion: karpenter.sh/v1
kind: Provisioner
metadata:
  name: default
spec:
  limits:
    resources:
      cpu: 20
  providerRef:
    apiVersion: karpenter.aws.k8s/v1
    kind: NodeTemplate
    name: default
  ttlSecondsAfterEmpty: 30
---
apiVersion: karpenter.aws.k8s/v1
kind: NodeTemplate
metadata:
  name: default
spec:
  amiFamily: AL2
  instanceProfile: KarpenterNodeInstanceProfile-karp-cluster
  subnetSelector:
    karpenter.sh/discovery: karp-cluster
  securityGroupSelector:
    karpenter.sh/discovery: karp-cluster
  userData:  |
    MIME-Version: 1.0
    Content-Type: multipart/mixed; boundary="//"
    --//
    Content-Type: text/x-shellscript; charset="us-ascii"
    #!/bin/bash
    echo "Running custom user data script"
    --//--      
```

Reply:
Agreed on what it would look like with v1alpha6, I think we're on the same page there. 

@bwagner5 and I spoke about the namespacing issue for a while. I don't think either of us believes Karpenter should have the ability to watch for ConfigMaps across namespaces and it's something we might scope down in the future. I don't think the default namespace makes sense either, but I think the namespace that karpenter components are running within seems like an okay choice to me and is something we'd always have access to watch on. I believe @tzneal has another initiative going to move some Karpenter config into a ConfigMap and we're probably going to rely on the karpenter namespace for that too. 

I can change it to `userDataConfigMap`, that sounds okay to me. I don't fully agree that inlining is better just because we push it to the end. I think having large amounts of TOML / Shell scripts embedded within YAML is just not great UX. and makes editing really tough.

Reply:
> I don't think the default namespace makes sense either, but I think the namespace that karpenter components are running within seems like an okay choice to me and is something we'd always have access to watch on. 

We default to installation to the karpenter namespace, but some customers prefer to install all system components into a different shared system namespace or even `kube-system`. In these cases, karpenter would have access to read config maps like `aws-auth`. 

> I believe @tzneal has another initiative going to move some Karpenter config into a ConfigMap and we're probably going to rely on the karpenter namespace for that too.

The difference is that the name of this configmap is well known vs granting access to all configmaps. In general, I think we're overly permissive with most of our RBAC rules. Really we should be aiming to only have permission to explicitly named objects that we manage (e.g. `config-logging`, `config-features`) and where necessary, we can have broader permissions (e.g. `pods`). In this example, Todd's new configmap is part of the karpenter installation, whereas what you're proposing is user configuration of karpenter. 

> I think having large amounts of TOML / Shell scripts embedded within YAML is just not great UX. 

Agreed -- but I think it's acceptable if we have line of site to a better long term solution within 2-3 months. 

> Agreed on what it would look like with v1alpha6, I think we're on the same page there.

Cool -- so we're just debating the best path to v1alpha6.

I think it's a more natural path to embed into the provisioner in the short term since the v1alpha6 change would simply be breaking out everything under `provider` into its own CRD (including userdata). Further, this avoids introducing a new concept to customers of what/how to configure configmaps that karpenter will read and then deprecating it a couple months later. Finally -- this also avoids adding RBAC scope (configmaps) to the karpenter controller, and then removing it later. RBAC changes need to be made judiciously due to their impact on the cluster's security posture.

Reply:
As discussed offline, moving to a new choice where the UserData is specified in an AWS Provider specific CRD.